### PR TITLE
Drools 5461 - Overall Transformations Workflow

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @mariofusco @jiripetrlik
see https://issues.redhat.com/browse/DROOLS-5461

Main modifications to this PR
1) implementing overall flow for transformations management
2) implemented overall DerivedFields management
3) implemented "Constant" and "FieldRef" Expressions
4) For drools models, moved "FieldType" map and "Types" list population to DroolsModelProvider

### CODE REVIEW

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/DerivedFieldFunctionUtils.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            return getExpressionMethodDeclaration(expression, arityCounter);
+        } else {
+            // TODO {gcardosi}
+            return new MethodDeclaration();
+        }
+    }
+
+    static MethodDeclaration getExpressionMethodDeclaration(final Expression expression, final AtomicInteger arityCounter) {
+        int methodArity = arityCounter.addAndGet(1);
+        if (expression instanceof Aggregate) {
```
Comment (gitgabrio) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/DerivedFieldFunctionUtils.java:89]: https://github.com/kiegroup/drools/pull/2962#discussion_r447713525

Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/DerivedFieldFunctionUtils.java:89]: @danielezonca 
I'm expecting not - those are the "default" Expressions. Extensions will be managed in future

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/factories/KiePMMLTreeModelASTFactory.java [hunk 2]
```diff
@@ -48,15 +49,26 @@ private KiePMMLTreeModelASTFactory() {
...
      */
-    public static KiePMMLDroolsAST getKiePMMLDroolsAST(DataDictionary dataDictionary, TreeModel model, final Map<String, KiePMMLOriginalTypeGeneratedType> fieldTypeMap) {
+    public static KiePMMLDroolsAST getKiePMMLDroolsAST(final DataDictionary dataDictionary,
+                                                       final TreeModel model,
+                                                       final Map<String, KiePMMLOriginalTypeGeneratedType> fieldTypeMap,
+                                                       final List<KiePMMLDroolsType> types) {
         logger.trace("getKiePMMLDroolsAST {} {}", dataDictionary, model);
         DATA_TYPE targetType = getTargetFieldType(dataDictionary, model);
-        List<KiePMMLDroolsType> types = KiePMMLDataDictionaryASTFactory.factory(fieldTypeMap).declareTypes(dataDictionary);
+//        List<KiePMMLDroolsType> types = fieldTypeMap.values()
```
Comment (gitgabrio) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-compiler/src/main/java/org/kie/pmml/models/drools/tree/compiler/factories/KiePMMLTreeModelASTFactory.java]: see https://github.com/kiegroup/drools/pull/2962#discussion_r448159866

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-tests/src/test/java/org/kie/pmml/regression/tests/LinearRegressionSampleWithTransformationsTest.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        final Map<String, Object> inputData = new HashMap<>();
+        inputData.put("age", age);
+        inputData.put("salary", salary);
+        inputData.put("car_location", car_location);
+        PMML4Result pmml4Result = evaluate(pmmlRuntime, inputData, MODEL_NAME);
+
+        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isNotNull();
+        Assertions.assertThat((double) pmml4Result.getResultVariables().get(TARGET_FIELD)).isCloseTo(expectedResult, TOLERANCE_PERCENTAGE);
+
+        // TODO {gcardosi} TO BE FIXED WITH DROOLS-5453
```
Comment (gitgabrio) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-tests/src/test/java/org/kie/pmml/regression/tests/LinearRegressionSampleWithTransformationsTest.java:90]: see https://github.com/kiegroup/drools/pull/2962#discussion_r448162251

Those are additional assertions - IMHO it does not make much sense to move them on a different test to only have the same computation made twice but verified with different assertions. BTW it is the way we have managed such situations until now

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-tests/src/test/java/org/kie/pmml/regression/tests/LogisticRegressionTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        inputData.put("variance", variance);
+        inputData.put("skewness", skewness);
+        inputData.put("curtosis", curtosis);
+        inputData.put("entropy", entropy);
+        PMML4Result pmml4Result = evaluate(pmmlRuntime, inputData, MODEL_NAME);
+
+        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isNotNull();
+        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isEqualTo(expectedResult);
+
+        // TODO {gcardosi} TO BE FIXED WITH DROOLS-5453
```
Comment (gitgabrio) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-regression/kie-pmml-models-regression-tests/src/test/java/org/kie/pmml/regression/tests/LogisticRegressionTest.java:93]: see https://github.com/kiegroup/drools/pull/2962#discussion_r448162165

See previous answer

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-tests/src/test/java/org/kie/pmml/models/drools/tree/tests/SampleMineTreeModelWithTransformationsTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    @Test
+    public void testSetPredicateTree() {
+        final Map<String, Object> inputData = new HashMap<>();
+        inputData.put("temperature", temperature);
+        inputData.put("humidity", humidity);
+        PMML4Result pmml4Result = evaluate(pmmlRuntime, inputData, MODEL_NAME);
+
+        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isNotNull();
+        Assertions.assertThat(pmml4Result.getResultVariables().get(TARGET_FIELD)).isEqualTo(expectedResult);
+        // // TODO {gcardosi} TO BE FIXED WITH DROOLS-5453
```
Comment (gitgabrio) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-tree/kie-pmml-models-drools-tree-tests/src/test/java/org/kie/pmml/models/drools/tree/tests/SampleMineTreeModelWithTransformationsTest.java:78]: see https://github.com/kiegroup/drools/pull/2962#discussion_r448163068

See above answer

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLASTTestUtils.java [hunk 6]
```diff
@@ -48,6 +54,19 @@ public static PredicateASTFactoryData getPredicateASTFactoryData(Predicate predi
         return new PredicateASTFactoryData(predicate, outputFields, rules, parentPath, currentRule, fieldTypeMap);
     }
 
+    public static Map<String, KiePMMLOriginalTypeGeneratedType> getFieldTypeMap(final DataDictionary dataDictionary, final TransformationDictionary transformationDictionary, final LocalTransformations localTransformations) {
+        final Map<String, KiePMMLOriginalTypeGeneratedType> toReturn = new HashMap<>();
```
Comment (gitgabrio) [inline kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLASTTestUtils.java:58]: see https://github.com/kiegroup/drools/pull/2962#discussion_r448164324

In general, this map is populated in different methods. For this specific case, the map has to be returned by that method

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/DerivedFieldFunctionUtils.java [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+import static org.kie.pmml.compiler.commons.utils.CommonCodegenUtils.getMethodDeclaration;
+import static org.kie.pmml.compiler.commons.utils.CommonCodegenUtils.getTypedClassOrInterfaceType;
+
+/**
+ * Class meant to provide <i>helper</i> methods to retrieve <code>Function</code> code-generators
+ * out of <code>DerivedField</code>s
+ */
+public class DerivedFieldFunctionUtils {
+
+    static final String KIEPMMLNAMEVALUE_LIST_PARAM = "param1"; // it is the first parameter
```
Comment (gitgabrio) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/DerivedFieldFunctionUtils.java:64]: see https://github.com/kiegroup/drools/pull/2962#discussion_r448188300

In the generated code there is a reference to that parameter. That reference must always be "param1".

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/DerivedFieldFunctionUtils.java [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+                                                   getDerivedFieldMethodDeclaration(derivedField, arityCounter)));
+        return toReturn;
+    }
+
+    static MethodDeclaration getDerivedFieldMethodDeclaration(final DerivedField derivedField, final AtomicInteger arityCounter) {
+        final Expression expression = derivedField.getExpression();
+        if (expression != null) {
+            return getExpressionMethodDeclaration(expression, arityCounter);
+        } else {
+            // TODO {gcardosi}
```
Comment (gitgabrio) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/main/java/org/kie/pmml/compiler/commons/utils/DerivedFieldFunctionUtils.java]: see https://github.com/kiegroup/drools/pull/2962#discussion_r448188702

**Review by jiripetrlik — APPROVED** [review]

File: kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLASTTestUtils.java (hunk 6 above)
Comment (danielezonca) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLASTTestUtils.java:58]: I'm still not sure, I do not expect a `factory` method just to populated one of the input. What about rename `factory` as `populate` / `init` / `fill`?

Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-common/src/test/java/org/kie/pmml/models/drools/utils/KiePMMLASTTestUtils.java:58]: Addressed (part of) in https://issues.redhat.com/browse/DROOLS-5272

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/enums/DATA_TYPE.java [hunk 9]
```diff
@@ -44,11 +44,19 @@
...
     TIME_SECONDS("timeSeconds", Long.class),
     DATE_TIME_SECONDS_SINCE_0("dateTimeSecondsSince[0]", Long.class),
     DATE_TIME_SECONDS_SINCE_1960("dateTimeSecondsSince[1960]", Long.class),
     DATE_TIME_SECONDS_SINCE_1970("dateTimeSecondsSince[1970]", Long.class),
-    DATE_TIME_SECONDS_SINCE_1980("dateTimeSecondsSince[1980]", Long.class);
+    DATE_TIME_SECONDS_SINCE_1980("dateTimeSecondsSince[1980]", Long.class),
+    DATE_TIME_SECONDS_SINCE_1990("x-dateTimeSecondsSince[1990]", Long.class),
+    DATE_TIME_SECONDS_SINCE_2000("x-dateTimeSecondsSince[2000]", Long.class),
+    DATE_TIME_SECONDS_SINCE_2010("x-dateTimeSecondsSince[2010]", Long.class),
+    DATE_TIME_SECONDS_SINCE_2020("x-dateTimeSecondsSince[2020]", Long.class);
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/enums/DATA_TYPE.java]: These are extensions, I think we can remove them. Wdyt?

Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/model/enums/DATA_TYPE.java]: @danielezonca 
Just done with last commit. This also solve the Sonar code smell

**Review by danielezonca — APPROVED** [review]

**Pull request merged**